Riley Estes
EE399 HW4
5/4/2023

Github link:
https://github.com/rileywe/Three-Layer-Neural-Networks

In [1]:
import numpy as np
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_squared_error
from sklearn.datasets import fetch_openml
from sklearn.decomposition import PCA
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from keras.models import Sequential
from keras.layers import LSTM, Dense

In [76]:
# Creates array data
X_nums=np.arange(0,31)
Y_nums=np.array([30, 35, 33, 32, 34, 37, 39, 38, 36, 36, 37, 39, 42, 45, 45, 41,
40, 39, 42, 44, 47, 49, 50, 49, 46, 48, 50, 53, 55, 54, 53])

# Imports MNIST data
mnist = fetch_openml('mnist_784', version=1)
X_mnist, Y_mnist = mnist.data, mnist.target

x:\anaconda\lib\site-packages\sklearn\datasets\_openml.py:932: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [77]:
# Initializes three layer feed forward neural network with a size 9 hidden layer
three_layer_NN = MLPRegressor(hidden_layer_sizes=(9,), activation='relu', solver='adam', max_iter=10000)

In [78]:
# Fits the given model to the training data
# Calculates the training and test mean squared errors as well as the least squared errors
# for each point in both datasets
def modelErrors (model, X_train, X_test, Y_train, Y_test):
    model.fit(X_train.reshape(-1, 1), Y_train)

    # LSE and MSE for training data
    Y_pred_train = model.predict(X_train.reshape(-1, 1))

    # Calculates LSE for each training data point
    LSE_train = (Y_train - Y_pred_train)**2

    # Prints LSE for each training data point
    for i in range(len(X_train)):
        print("Training data point {}: LSE = {:.2f}".format(i, LSE_train[i]))

    mse_train = mean_squared_error(Y_train, Y_pred_train)
    print("MSE on training data: {:.2f}".format(mse_train))

    print()

    # LSE and MSE for testing data
    Y_pred_test = model.predict(X_test.reshape(-1, 1))

    # Calculates LSE for each testing data point
    LSE_test = (Y_test - Y_pred_test)**2

    # Prints LSE for each testing data point
    for i in range(len(X_test)):
        print("Testing data point {}: LSE = {:.2f}".format(i, LSE_test[i]))

    mse_test = mean_squared_error(Y_test, Y_pred_test)
    print("MSE on test data: {:.2f}".format(mse_test))

In [85]:
# Fits the model using the first 20 points for training
X_train, X_test = X_nums[:20], X_nums[20:]
Y_train, Y_test = Y_nums[:20], Y_nums[20:]

three_layer_NN.fit(X_train.reshape(-1, 1), Y_train)

modelErrors(three_layer_NN, X_train, X_test, Y_train, Y_test)

Training data point 0: LSE = 4.16
Training data point 1: LSE = 5.05
Training data point 2: LSE = 0.22
Training data point 3: LSE = 4.40
Training data point 4: LSE = 0.52
Training data point 5: LSE = 2.75
Training data point 6: LSE = 9.23
Training data point 7: LSE = 2.01
Training data point 8: LSE = 1.45
Training data point 9: LSE = 3.33
Training data point 10: LSE = 2.09
Training data point 11: LSE = 0.00
Training data point 12: LSE = 5.34
Training data point 13: LSE = 21.99
Training data point 14: LSE = 16.55
Training data point 15: LSE = 0.31
Training data point 16: LSE = 4.73
Training data point 17: LSE = 14.41
Training data point 18: LSE = 2.01
Training data point 19: LSE = 0.00
MSE on training data: 5.03

Testing data point 0: LSE = 2.64
Testing data point 1: LSE = 5.11
Testing data point 2: LSE = 3.55
Testing data point 3: LSE = 0.24
Testing data point 4: LSE = 23.61
Testing data point 5: LSE = 17.90
Testing data point 6: LSE = 12.98
Testing data point 7: LSE = 3.90
Testing data

In [86]:
# Fits the model using the first and last 10 points as training
X_train = np.concatenate((X_nums[:10], X_nums[-10:]))
Y_train = np.concatenate((Y_nums[:10], Y_nums[-10:]))
X_test = X_nums[10:20]
Y_test = Y_nums[10:20]

three_layer_NN.fit(X_train.reshape(-1, 1), Y_train)

modelErrors(three_layer_NN, X_train, X_test, Y_train, Y_test)

Training data point 0: LSE = 0.00
Training data point 1: LSE = 14.17
Training data point 2: LSE = 0.23
Training data point 3: LSE = 3.24
Training data point 4: LSE = 1.17
Training data point 5: LSE = 1.37
Training data point 6: LSE = 5.99
Training data point 7: LSE = 0.53
Training data point 8: LSE = 3.98
Training data point 9: LSE = 7.38
Training data point 10: LSE = 2.66
Training data point 11: LSE = 3.64
Training data point 12: LSE = 0.04
Training data point 13: LSE = 12.49
Training data point 14: LSE = 5.09
Training data point 15: LSE = 0.95
Training data point 16: LSE = 1.70
Training data point 17: LSE = 6.66
Training data point 18: LSE = 0.74
Training data point 19: LSE = 0.74
MSE on training data: 3.64

Testing data point 0: LSE = 5.94
Testing data point 1: LSE = 1.34
Testing data point 2: LSE = 1.25
Testing data point 3: LSE = 11.55
Testing data point 4: LSE = 7.17
Testing data point 5: LSE = 4.18
Testing data point 6: LSE = 14.17
Testing data point 7: LSE = 30.09
Testing data 

In [87]:
# Fits the model to the test data in the middle
three_layer_NN.fit(X_test.reshape(-1, 1), Y_test)

modelErrors(three_layer_NN, X_train, X_test, Y_train, Y_test)

Training data point 0: LSE = 0.05
Training data point 1: LSE = 6.09
Training data point 2: LSE = 0.16
Training data point 3: LSE = 4.57
Training data point 4: LSE = 0.76
Training data point 5: LSE = 1.95
Training data point 6: LSE = 7.09
Training data point 7: LSE = 0.86
Training data point 8: LSE = 3.25
Training data point 9: LSE = 6.43
Training data point 10: LSE = 2.79
Training data point 11: LSE = 3.75
Training data point 12: LSE = 0.04
Training data point 13: LSE = 12.46
Training data point 14: LSE = 5.12
Training data point 15: LSE = 0.99
Training data point 16: LSE = 1.62
Training data point 17: LSE = 6.45
Training data point 18: LSE = 0.65
Training data point 19: LSE = 0.86
MSE on training data: 3.30

Testing data point 0: LSE = 5.15
Testing data point 1: LSE = 1.00
Testing data point 2: LSE = 1.60
Testing data point 3: LSE = 12.48
Testing data point 4: LSE = 7.84
Testing data point 5: LSE = 3.74
Testing data point 6: LSE = 13.44
Testing data point 7: LSE = 29.15
Testing data p

Compare the models fit in homework one to the neural network:
These models both perform worse on the data set than the simpler linear and sinusoidal regression models. 
For the first dataset where the first 20 points are used for training, the test MSE is 11.15 for the neural network, and 3.53 for the linear model. For the second dataset where the first and last 10 points are used for training, the test MSE is 8.97 for the neural network, and 2.95 for the linear model. When fitted to the test data instead, the training MSE for the neural network is 3.30, which is comparable to the linear fit, except that the test MSE is still very bad. The linear regression model performs much better in both cases than the neural network at a fraction of the computational cost, making it the obvious choice for simple curve-fitting tasks such as these. 

In [63]:
# Computes the first 20 PCA modes of the MNIST data
pca = PCA(n_components=20)
X_pca = pca.fit_transform(X_mnist)

# Splits the data into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X_pca, Y_mnist, test_size=0.2, random_state=0)

In [65]:
# Trains a 3 layer feedforward neural network on the PCA data
three_layer_mnist = MLPClassifier(hidden_layer_sizes=(100,), activation='relu', solver='adam', max_iter=1000, random_state=0)
three_layer_mnist.fit(X_train, Y_train)

# Evaluates the performance of the model on the test data
accuracy = three_layer_mnist.score(X_test, Y_test)*100
print("Feed forward NN test accuracy: {:.2f}%".format(accuracy))

Feed forward NN test accuracy: 95.69%


In [69]:
# Trains and evaluates a 3 layer LSTM with a size 100 hidden layer, 10 epochs, and batch size of 128 on the MNIST data
model_lstm = Sequential()
model_lstm.add(LSTM(128, input_shape=(X_train.shape[1], 1)))
model_lstm.add(Dense(10, activation='softmax'))
model_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
X_train_lstm = X_train.reshape((X_train.shape[0], X_train.shape[1], 1))
model_lstm.fit(X_train_lstm, np.eye(10)[Y_train.astype(int)], epochs=10, batch_size=128)
_, acc_lstm = model_lstm.evaluate(X_test.reshape((X_test.shape[0], X_test.shape[1], 1)), np.eye(10)[Y_test.astype(int)], verbose=0)
print("LSTM test accuracy: {:.2f}%".format(acc_lstm*100))

Epoch 1/10
438/438 [==============================] - 14s 27ms/step - loss: 1.9329 - accuracy: 0.3121
Epoch 2/10
438/438 [==============================] - 12s 28ms/step - loss: 1.2290 - accuracy: 0.5823
Epoch 3/10
438/438 [==============================] - 11s 24ms/step - loss: 0.8600 - accuracy: 0.7159
Epoch 4/10
438/438 [==============================] - 10s 22ms/step - loss: 0.6710 - accuracy: 0.7827
Epoch 5/10
438/438 [==============================] - 10s 23ms/step - loss: 0.5811 - accuracy: 0.8129
Epoch 6/10
438/438 [==============================] - 10s 23ms/step - loss: 0.5234 - accuracy: 0.8342
Epoch 7/10
438/438 [==============================] - 10s 23ms/step - loss: 0.4726 - accuracy: 0.8515
Epoch 8/10
438/438 [==============================] - 10s 23ms/step - loss: 0.4387 - accuracy: 0.8618
Epoch 9/10
438/438 [==============================] - 10s 22ms/step - loss: 0.4180 - accuracy: 0.8691
Epoch 10/10
438/438 [==============================] - 10s 24ms/step - loss: 0.377

In [67]:
# Trains and evaluates an SVM on the MNIST data
model_svm = SVC()
model_svm.fit(X_train, Y_train)
acc_svm = model_svm.score(X_test, Y_test)*100
print("SVM test accuracy: {:.2f}%".format(acc_svm))

SVM test accuracy: 97.30%


In [68]:
# Trains and evaluates a decision tree on the MNIST data
model_dt = DecisionTreeClassifier()
model_dt.fit(X_train, Y_train)
acc_dt = model_dt.score(X_test, Y_test)*100
print("Decision tree test accuracy: {:.2f}%".format(acc_dt))

Decision tree test accuracy: 84.41%
